library/core/src/gc.rs

```diff
@@ -21,3 +21,18 @@ pub fn gc_layout<T>() -> (u64, u64) {
     let layout = intrinsics::gc_layout::<T>();
     (layout[0], layout[1])
 }
+
+#[unstable(feature = "gc", issue = "none")]
+#[cfg_attr(not(bootstrap), lang = "notrace")]
+pub auto trait NoTrace {}
+
+impl !NoTrace for usize {}
+
+#[cfg(target_pointer_width = "64")]
+impl !NoTrace for u64 {}
+
+#[cfg(target_pointer_width = "32")]
+impl !NoTrace for u32 {}
+
+impl<T: ?Sized> !NoTrace for *mut T {}
+impl<T: ?Sized> !NoTrace for *const T {}
```

Comment:
Do we have to do anything for smart pointer types?

Reply:
We'll want to negatively implement `Gc<T>` but this isn't defined in `rustc_boehm` so that will happen in a separate repo.

Reply:
I guess Edd's also thinking of `Rc` and the like?

Reply:
Ah I see. No I don't think so, `Rc` and the like will be _poisoned_ automatically if its contents is `!NoTrace`.  What we really care about here are raw pointers.